ampd/src/solana/mod.rs

```diff
@@ -89,47 +93,56 @@ where
         }
     };
 
-    // pare the events
-    let event_stack = gateway_event_stack::build_program_event_stack(
-        match_context,
-        logs,
-        gateway_event_stack::parse_gateway_logs,
-    );
-
-    for invocation_state in event_stack {
-        use gateway_event_stack::ProgramInvocationState::*;
-        let (vote, gateway_events) = match invocation_state {
-            Succeeded(events) => (
-                {
-                    // if tx was successful and ix invocation was successful,
-                    // then the final outcome (if event can be found) will be of `Vote::SucceededOnChain`
-                    if tx.err.is_none() {
-                        Vote::SucceededOnChain
-                    } else {
-                        // if tx was NOT successful then we don't care if the ix invocatoin succeeded,
-                        // therefore the final outcome (if event can be found) will be of `Vote::FailedOnChain`
-                        Vote::FailedOnChain
-                    }
-                },
-                events,
-            ),
-            Failed(events) | InProgress(events) => (Vote::FailedOnChain, events),
-        };
-
-        if let Some((_, event)) = gateway_events
-            .into_iter()
-            .find(|(idx, _)| *idx == desired_event_idx)
-        {
-            if events_are_equal(&event) {
-                // proxy the desired vote status of whether the ix succeeded
-                return vote;
-            }
+    // Check in the logs in a backward way the invocation comes from the gateway
+    if !event_comes_from_gateway(logs, desired_event_idx) {
+        error!("Event does not come from the gateway");
+        return Vote::NotFound;
+    }
+    // Check that the event index is in the logs
+    let Some(event_log) = logs.get(desired_event_idx) else {
+        error!("Event index is out of bounds");
+        return Vote::NotFound;
+    };
 
-            warn!(?event, "event was found, but contents were not equal");
+    // Second ensure we can parse the event
+    let event = match gateway_event_stack::parse_gateway_logs(event_log) {
+        Ok(ev) => ev,
+        Err(err) => {
+            error!("Cannot parse the gateway log: {}", err);
             return Vote::NotFound;
         }
+    };
+
+    if events_are_equal(&event) {
+        Vote::SucceededOnChain
+    } else {
+        warn!(?event, "event was found, but contents were not equal");
+        Vote::NotFound
     }
+}
 
-    warn!("not found");
-    Vote::NotFound
+// Check backward in the logs if the invocation comes from the gateway program,
+// skipping native program invocations
+fn event_comes_from_gateway(logs: &[String], desired_event_idx: usize) -> bool {
```

Comment:
Probably should add an example of how the logs look like, why we are parsing "upwards" from the idx, why don't we check for "failed" state (the tx success needs to be checked beforehand manually), and why do we skip "system program id"

Reply:
Added in last optimisations commit at fb5d493112711725754aef870b3d1b418a177422